app/src/main/java/org/mifos/mobile/viewModels/ReviewLoanApplicationViewModel.kt

```diff
@@ -63,14 +67,24 @@ class ReviewLoanApplicationViewModel @Inject constructor(var dataManager: DataMa
 
     fun getLoanState() = loanState
 
-    fun submitLoan(): Observable<ResponseBody?>? {
+    private val _submitLoanUiState = MutableLiveData<SubmitLoanUiState>()
+    val submitLoanUiState: LiveData<SubmitLoanUiState> = _submitLoanUiState
+
+    fun submitLoan() = viewModelScope.launch(Dispatchers.IO) {
         loansPayload.productName = null
         loansPayload.loanPurpose = null
         loansPayload.currency = null
-        return if (loanState == LoanState.CREATE) {
-            dataManager?.createLoansAccount(loansPayload)
-        } else {
-            dataManager?.updateLoanAccount(loanId, loansPayload)
+
+        _submitLoanUiState.postValue(SubmitLoanUiState.Loading)
+        try {
+            val response = if (loanState == LoanState.CREATE) {
+                dataManager?.createLoansAccount(loansPayload)
```

Comment:
viewmodel should take data from the repository and not directly from the source. Make an interface by the name of `ReviewLoanApplicationRepository` and then implement this interface in a class that you can call as `ReviewLoanApplicationRepositoryImp`  .  You can take reference from the code that we wrote during GSoC.